evm/src/fuzz/invariant/executor.rs

```diff
@@ -0,0 +1,420 @@
+                        invariant_contract.invariant_functions.len()
+                    {
+                        return Err(TestCaseError::fail("All invariants have been broken."))
+                    }
+                }
+
+                // Before each run, we must reset the backend state.
+                let mut executor = blank_executor.borrow().clone();
+
+                // Used for stat reports (eg. gas usage).
+                let mut fuzz_runs = vec![];
+
+                // Created contracts during a run.
+                let mut created_contracts = vec![];
+
+                'fuzz_run: for _ in 0..test_options.depth {
+                    let (sender, (address, calldata)) =
+                        inputs.last().expect("to have the next randomly generated input.");
+
+                    // Executes the call from the randomly generated sequence.
+                    let call_result = executor
+                        .call_raw(*sender, *address, calldata.0.clone(), U256::zero())
+                        .expect("could not make raw evm call");
+
+                    // Collect data for fuzzing from the state changeset.
+                    let state_changeset =
+                        call_result.state_changeset.to_owned().expect("to have a state changeset.");
+
+                    collect_state_from_call(
+                        &call_result.logs,
+                        &state_changeset,
+                        fuzz_state.clone(),
+                    );
+                    collect_created_contracts(
+                        &state_changeset,
+                        self.project_contracts,
+                        self.setup_contracts,
+                        targeted_contracts.clone(),
+                        &mut created_contracts,
+                    );
+
+                    // Commit changes to the database.
+                    executor.backend_mut().db.commit(state_changeset);
+
+                    fuzz_runs.push(FuzzCase {
+                        calldata: calldata.clone(),
+                        gas: call_result.gas,
+                        stipend: call_result.stipend,
+                    });
+
+                    if !can_continue(
+                        &invariant_contract,
+                        call_result,
+                        &executor,
+                        &inputs,
+                        &mut failures.borrow_mut(),
+                        test_options,
+                    ) {
+                        break 'fuzz_run
+                    }
+
+                    // Generates the next call from the run using the recently updated
+                    // dictionary.
+                    inputs.extend(
+                        strat
+                            .new_tree(&mut branch_runner.borrow_mut())
+                            .map_err(|_| TestCaseError::Fail("Could not generate case".into()))?
+                            .current(),
+                    );
```

Comment:
does this allow for re-using the same input? e.g. call `deposit(100)` and then `deposit(100)` again?

Reply:
If chance allows it, yes